`<chrono>`: `format()` should accept `%X` and `%EX` for `duration` and `hh_mm_ss`

Found by the upcoming libc++ test update. Accepted by libc++ and libstdc++: https://godbolt.org/z/oWGKosE3d

```
D:\GitHub\STL\out\x64>type meow.cpp
```
```cpp
#include <cassert>
#include <chrono>
#include <format>
#include <string>
using namespace std;

int main() {
    // N4964 [time.format]/6: "However, if a flag refers to a "time of day" (e.g., `%H`, `%I`, `%p`, etc.),
    // then a specialization of `duration` is interpreted as the time of day elapsed since midnight."
    // [tab:time.format.spec]: "`%X` The locale's time representation.
    // The modified command `%EX` produces the locale's alternate time representation."

    // libcxx/test/std/time/time.syn/formatter.duration.pass.cpp
    assert(format("{:%X}", 0s) == "00:00:00");
    assert(format("{:%EX}", 0s) == "00:00:00");

    // libcxx/test/std/time/time.syn/formatter.file_time.pass.cpp
    // libcxx/test/std/time/time.syn/formatter.hh_mm_ss.pass.cpp
    // libcxx/test/std/time/time.syn/formatter.local_time.pass.cpp
    // libcxx/test/std/time/time.syn/formatter.sys_time.pass.cpp
    assert(format("{:%X}", chrono::hh_mm_ss{0s}) == "00:00:00");
    assert(format("{:%EX}", chrono::hh_mm_ss{0s}) == "00:00:00");
}
```
```
D:\GitHub\STL\out\x64>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od meow.cpp && meow
meow.cpp
meow.cpp(14): error C7595: 'std::basic_format_string<char,std::chrono::duration<__int64,std::ratio<1,1>>>::basic_format_string': call to immediate function is not a constant expression
D:\GitHub\STL\out\x64\out\inc\chrono(5321): note: failure was caused by call of undefined function or one not declared 'constexpr'
D:\GitHub\STL\out\x64\out\inc\chrono(5321): note: see usage of 'std::_Throw_format_error'
[...]

D:\GitHub\STL\out\x64>clang-cl /EHsc /nologo /W4 /std:c++latest /MTd /Od meow.cpp && meow
meow.cpp(14,19): error: call to consteval function 'std::basic_format_string<char, std::chrono::duration<long
      long>>::basic_format_string<char[6]>' is not a constant expression
    assert(format("{:%X}", 0s) == "00:00:00");
                  ^
D:\GitHub\STL\out\x64\out\inc\chrono(5321,21): note: non-constexpr function '_Throw_format_error' cannot be used in a
      constant expression
                    _Throw_format_error("Invalid type.");
                    ^
[...]
```


For `duration`, we accept a superset of what `hh_mm_ss<seconds>` accepts:
https://github.com/microsoft/STL/blob/0403d19f5461fd15983737c3f01ec34800ea9275/stl/inc/chrono#L5378-L5379

But `hh_mm_ss` doesn't accept `'X'`:
https://github.com/microsoft/STL/blob/0403d19f5461fd15983737c3f01ec34800ea9275/stl/inc/chrono#L5399-L5401

--- a/stl/inc/chrono
+++ b/stl/inc/chrono
@@ -5398,7 +5398,7 @@ namespace chrono {
                     || _Is_valid_type<month>(_Type) || _Is_valid_type<day>(_Type) || _Is_valid_type<weekday>(_Type);
             } else if constexpr (_Is_specialization_v<_Ty, hh_mm_ss>) {
                 return _Type == 'H' || _Type == 'I' || _Type == 'M' || _Type == 'S' || _Type == 'r' || _Type == 'R'
-                    || _Type == 'T' || _Type == 'p';
+                    || _Type == 'T' || _Type == 'p' || _Type == 'X';
             } else if constexpr (_Is_any_of_v<_Ty, sys_info, local_info>) {
                 return _Type == 'z' || _Type == 'Z';
             } else if constexpr (_Is_specialization_v<_Ty, time_point>) {
@@ -5407,7 +5407,7 @@ namespace chrono {
                         return true;
                     }
                 }
-                return _Type == 'c' || _Type == 'X' || _Is_valid_type<year_month_day>(_Type)
+                return _Type == 'c' || _Is_valid_type<year_month_day>(_Type)
                     || _Is_valid_type<hh_mm_ss<seconds>>(_Type);
             } else if constexpr (_Is_specialization_v<_Ty, _Local_time_format_t>) {
                 return _Type == 'z' || _Type == 'Z' || _Is_valid_type<decltype(_Ty::_Time)>(_Type);

--- a/tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
+++ b/tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
@@ -282,6 +282,10 @@ void test_duration_formatter() {
     assert(format(STR("{:%T %j}"), -days{4} - 23h - 30min) == STR("-119:30:00 4"));
     assert(format(STR("{:%T %j}"), duration<float, days::period>{1.55f}) == STR("37:11:59 1"));
     assert(format(STR("{:%T %j}"), duration<float, days::period>{-1.55f}) == STR("-37:11:59 1"));
+
+    // GH-4247: <chrono>: format() should accept %X and %EX for duration and hh_mm_ss
+    assert(format(STR("{:%X}"), 9h + 7min + 5s) == STR("09:07:05"));
+    assert(format(STR("{:%EX}"), 9h + 7min + 5s) == STR("09:07:05"));
 }
 
 template <typename CharT>
@@ -765,6 +769,10 @@ void test_hh_mm_ss_formatter() {
     assert(format(STR("{:%H}"), hh_mm_ss{24h}) == STR("24"));
     assert(format(STR("{:%H}"), hh_mm_ss{-24h}) == STR("-24"));
     assert(format(STR("{:%M %S}"), hh_mm_ss{27h + 12min + 30s}) == STR("12 30"));
+
+    // GH-4247: <chrono>: format() should accept %X and %EX for duration and hh_mm_ss
+    assert(format(STR("{:%X}"), hh_mm_ss{9h + 7min + 5s}) == STR("09:07:05"));
+    assert(format(STR("{:%EX}"), hh_mm_ss{9h + 7min + 5s}) == STR("09:07:05"));
 }
 
 void test_exception_classes() {

